// @alilc/lowcode-utils/lib/is-react.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.acceptsRef = acceptsRef;
exports.isReactClass = isReactClass;
exports.isReactComponent = isReactComponent;
exports.wrapReactClass = wrapReactClass;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = require("react");
var _cloneEnumerableProperty = require("./clone-enumerable-property");
var _excluded = ["children"];
var hasSymbol = typeof Symbol === 'function' && Symbol["for"];
var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol["for"]('react.forward_ref') : 0xead0;
function isReactClass(obj) {
  return obj && obj.prototype && (obj.prototype.isReactComponent || obj.prototype instanceof _react.Component);
}
function acceptsRef(obj) {
  var _obj$prototype;
  return (obj === null || obj === void 0 ? void 0 : (_obj$prototype = obj.prototype) === null || _obj$prototype === void 0 ? void 0 : _obj$prototype.isReactComponent) || obj.$$typeof && obj.$$typeof === REACT_FORWARD_REF_TYPE;
}
function isForwardRefType(obj) {
  return (obj === null || obj === void 0 ? void 0 : obj.$$typeof) && (obj === null || obj === void 0 ? void 0 : obj.$$typeof) === REACT_FORWARD_REF_TYPE;
}
function isReactComponent(obj) {
  return obj && (isReactClass(obj) || typeof obj === 'function' || isForwardRefType(obj));
}
function wrapReactClass(view) {
  var ViewComponentClass = /*#__PURE__*/function (_Component) {
    (0, _inheritsLoose2["default"])(ViewComponentClass, _Component);
    function ViewComponentClass() {
      return _Component.apply(this, arguments) || this;
    }
    var _proto = ViewComponentClass.prototype;
    _proto.render = function render() {
      var _this$props = this.props,
        children = _this$props.children,
        other = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, _excluded);
      return /*#__PURE__*/(0, _react.createElement)(view, other, children);
    };
    return ViewComponentClass;
  }(_react.Component);
  ViewComponentClass = (0, _cloneEnumerableProperty.cloneEnumerableProperty)(ViewComponentClass, view);
  ViewComponentClass.displayName = view.displayName;
  return ViewComponentClass;
}
