// @alilc/lowcode-plugin-code-editor/lib/components/ErrorBoundary/ErrorBoundary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.ErrorBoundary = void 0;
var _message = _interopRequireDefault(require("@alifd/next/lib/message"));
var _button = _interopRequireDefault(require("@alifd/next/lib/button"));
var _loading = _interopRequireDefault(require("@alifd/next/lib/loading"));
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var _ErrorBoundary = _interopRequireDefault(require("./ErrorBoundary.less"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var ErrorBoundary = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(ErrorBoundary, _Component);
  function ErrorBoundary() {
    var _this;
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      hasError: false,
      info: null
    };
    _this._handleReset = function () {
      _this.setState({
        hasError: false
      });
    };
    return _this;
  }
  ErrorBoundary.getDerivedStateFromError = function getDerivedStateFromError() {
    return {
      hasError: true
    };
  };
  var _proto = ErrorBoundary.prototype;
  _proto.componentDidCatch = function componentDidCatch(error, info) {
    var onCatch = this.props.onCatch;
    if (onCatch) {
      onCatch(error, info);
      this.setState({
        info: info
      });
    }
  };
  _proto.render = function render() {
    var children = this.props.children;
    var _this$state = this.state,
      hasError = _this$state.hasError,
      info = _this$state.info;
    // const error = new Error('component render error');
    var errorInfo = {
      stack: (info === null || info === void 0 ? void 0 : info.componentStack) || ''
    };
    if (!hasError) {
      return /*#__PURE__*/_react["default"].createElement(_react.Suspense, {
        fallback: /*#__PURE__*/_react["default"].createElement(_loading["default"], {
          visible: true
        })
      }, children);
    }
    return /*#__PURE__*/_react["default"].createElement(_message["default"], {
      title: "\u51FA\u9519\u4E86~",
      type: "error"
    }, /*#__PURE__*/_react["default"].createElement("p", null, "\u8BE6\u7EC6\u9519\u8BEF: ", errorInfo || '未知原因'), /*#__PURE__*/_react["default"].createElement("div", {
      className: _ErrorBoundary["default"].actions
    }, /*#__PURE__*/_react["default"].createElement(_button["default"], {
      onClick: this._handleReset,
      size: "small"
    }, "\u91CD\u8BD5")));
  };
  return ErrorBoundary;
}(_react.Component);
exports.ErrorBoundary = ErrorBoundary;
